application-admintools-default/src/main/java/com/xwiki/admintools/internal/uploadJob/UploadJob.java

```diff
@@ -0,0 +1,223 @@
+                    Files.copy(backupFile.toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    boolean backupFileDeleteFailure = !handleFileDeletion(backupFile, FILE_TYPE_BACKUP);
+                    if (newFileDeleteFailure || backupFileDeleteFailure) {
+                        continue;
+                    }
+                } catch (IOException e) {
+                    backupFailureMessage("copy", backupFile.getName(), targetFile.getName());
+                    continue;
+                }
+            }
+            JobResult log = new JobResult("adminTools.jobs.upload.batch.backup.success", JobResultLevel.INFO,
+                jobResource.getNewFileName());
+            status.addLog(log);
+            progressManager.endStep(this);
+            Thread.yield();
+        }
+    }
+
+    private boolean handleFileDeletion(File file, String fileType)
+    {
+        if (file != null) {
+            try {
+                Files.deleteIfExists(file.toPath());
+            } catch (IOException e) {
+                backupFailureMessage(fileType, file.getName());
+                progressManager.endStep(this);
+                Thread.yield();
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void backupFailureMessage(String translation, String... parameters)
+    {
+        String translationHint = String.format("adminTools.jobs.upload.batch.backup.%s.fail", translation);
+        JobResult log = new JobResult(translationHint, JobResultLevel.ERROR, parameters);
+        status.addLog(log);
+    }
+
+    private void batchSave() throws IOException
+    {
+        for (UploadPackageJobResource jobResource : jobResourceList) {
+            progressManager.startStep(this);
+            File targetFile = jobResource.getTargetFile();
+            if (targetFile.exists()) {
+                String filePath = targetFile.getParent();
+                if (!targetFile.delete()) {
+                    status.addLog(new JobResult("adminTools.jobs.upload.batch.save.fail", JobResultLevel.ERROR,
+                        targetFile.getName()));
+                    throw new RuntimeException(
+                        String.format("Failed to delete original file [%s].", targetFile.getName()));
+                }
+                File newFile = new File(String.format(NEW_FILE_PATH_FORMAT, filePath, jobResource.getNewFileName()));
+                Files.write(newFile.toPath(), jobResource.getNewFileContent().toByteArray());
+            } else {
+                Files.write(targetFile.toPath(), jobResource.getNewFileContent().toByteArray());
+            }
+            jobResource.setUploaded();
+            status.addLog(new JobResult("adminTools.jobs.upload.batch.save.success", JobResultLevel.INFO,
+                jobResource.getNewFileName()));
+            progressManager.endStep(this);
+            Thread.yield();
+        }
+    }
+}
```

Comment:
All these additional code lines makes the file convoluted and harder to follow. I'll take another look on this after you change the logging process as suggested. 

As a note, you should log more information for this job. Like "starting the batch save", "failed batch save. Starting batch restore". Things like these.